### PR TITLE
Overhaul thrown exceptions

### LINKED ISSUE #189: WP_Mock should define its own exception objects
## Feature request

WP_Mock should define its own base exception, e.g. `WP_MockException` to use in place of generic `Exception` throwables. 

## Proposed solution

Add a new `WP_Mock\WP_MockException` in `WP_Mock/Exceptions` that extends `Exception`. No further changes needed for now. As we overhaul more WP_Mock code we should throw this exception and create more specific exceptions that extend the base exception where applicable. Specific exceptions could be useful to identify specific problems with the proejct.

### PR DESCRIPTION
**Pull request by unfulvio-godaddy**
# Summary 

~~Adds a `WP_MockException` object and overhauls exceptions used or referenced within project.~~

After consideration, I decided to go for more specific exceptions in the PHP stdlib - such as `RunTimeException` or `InvalidArgumentException` when applicable. I realized that WP_Mock already uses other domain-specific exceptions when  it comes to assertions failed or mockery exceptions, so perhaps at the moment we don't need a specific WP_Mock exception.

Had to make additional changes due to PhpStan errors after updating code.

## Contributor checklist 

- [x] I agree to follow this project's [**Code of Conduct**](https://github.com/10up/.github/blob/trunk/CODE_OF_CONDUCT.md).
- [x] I have updated the documentation accordingly 
- [x] I have added tests to cover changes introduced by this pull request
- [x] All new and existing tests pass

## Testing 

- [x] Tests and code analysis passes

### Reviewer checklist 

 

- [x] Code changes review
- [ ] Documentation changes review
- [x] Unit tests pass
- [x] Static analysis passes

### CODE REVIEW

File: php/WP_Mock/Exceptions/WP_MockException.php [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+namespace WP_Mock\Exceptions;
+
+use Exception;
+use Throwable;
+
+/**
+ * Base WP_Mock Exception.
+ */
+class WP_MockException extends Exception
```
Comment (agibson-godaddy) [inline php/WP_Mock/Exceptions/WP_MockException.php]: Just a question: why the `WP_Mock` prefix? It's already in a `WP_Mock` namespace, so it seems unnecessary. Why not something like `WP_Mock\Exceptions\BaseException` ?

File: php/WP_Mock/Exceptions/WP_MockException.php [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    /** @var int default from {@see Exception} - child implementations of this class may override this */
+    protected $code = 0;
+
+    /**
+     * Exception constructor.
+     *
+     * @param string $message
+     * @param Throwable|null $previous
+     */
+    public function __construct(string $message = '', Throwable $previous = null)
```
Comment (agibson-godaddy) [inline php/WP_Mock/Exceptions/WP_MockException.php]: Why not include the code in the constructor? (I did see your comment about child implementations allowing for overrides, but just curious why not put it in the constructor anyway.)

**Review by agibson-godaddy — DISMISSED** [review]
Approving this because it seems fine, but I did leave a couple comments for you to consider.

File: php/WP_Mock/Exceptions/WP_MockException.php (hunk 2 above)
Comment (unfulvio-godaddy) [reply php/WP_Mock/Exceptions/WP_MockException.php]: I checked our usage in MWC packages we also seem to put the code directly in the exception property - we also tend to use codes mapped to HTTP errors sometimes, which I doubt is going to happen in WP_Mock.

File: php/WP_Mock/Exceptions/WP_MockException.php (hunk 1 above)
Comment (unfulvio-godaddy) [reply php/WP_Mock/Exceptions/WP_MockException.php]: Good point, maybe I'll just rename that to `Exception` in the `WP_Mock` namespace; more specific exceptions can define better names

**Comment by unfulvio-godaddy** [comment]
@agibson-godaddy sorry I went to a different direction after reading your comments but mostly re-evaluating the purpose of this PR - in the end maybe WP_Mock is already using some specific exceptions depending on context (from PhpUnit expectation failed exception, mockery exceptions, etc.). I have added a couple of `InvalidArgumentException` or `RunTimeException` where it seemed to make most sense. Perhaps there is no need to use a generic exception at the moment.

**Review by agibson-godaddy — APPROVED** [review]

**Pull request merged**